src/txn/txn_timestamp.c

```diff
@@ -452,29 +453,34 @@ __txn_assert_after_reads(WT_SESSION_IMPL *session, const char *op, wt_timestamp_
 
     txn_global = &S2C(session)->txn_global;
 
-    __wt_readlock(session, &txn_global->rwlock);
-    /* Walk the array of concurrent transactions. */
     WT_ORDERED_READ(session_cnt, S2C(session)->session_cnt);
     WT_STAT_CONN_INCR(session, txn_walk_sessions);
+    WT_STAT_CONN_INCRV(session, txn_sessions_walked, session_cnt);
+
+    __wt_readlock(session, &txn_global->rwlock);
```

Comment:
Curious about this movement of the read-lock. Was the prior code a correctness problem? Are you sure it's correct to get read in the `session_cnt` before having a lock?

Reply:
Interesting. The motivation makes sense, but the correctness concern wasn't the answer I was expecting! Do you mind expanding a bit more :)

I know the underlying sessions array is static piece of memory. I'm more interested that this function can be reading values from WT_TXN_SHARED structures that are being modified. Maybe this is a case that there's agreement the code is (strictly speaking) a data race, but maybe that race only manifests as false negatives? That said, I don't know what the `txn_global->rwlock` is guarding.

One example: can the session_cnt increment after the read but before taking the lock? Can this scenario result in us not observing a violation because we stopped checking one session too soon? Of course it's plausible with the lock in its original location that we acquire it before the new session comes into existence. And we've deterministically removed it from being checked.

Alternatively if the session_cnt decrements (I think that can also happen?), we'll be checking a session that's not longer active. Is it the case that all sessions outside of the valid "session_cnt" range are appropriately "zeroed out" and thus won't trip this check?

This last bullet point I think gets back what the `txn_global->rwlock` protects. While we do have that lock in the "reader" mode, is it guaranteed sessions cannot acquire new snapshots nor release existing snapshots? Can you confirm that's the case?

Reply:
@dgottlieb

> I know the underlying sessions array is static piece of memory. I'm more interested that this function can be reading values from WT_TXN_SHARED structures that are being modified. Maybe this is a case that there's agreement the code is (strictly speaking) a data race, but maybe that race only manifests as false negatives? That said, I don't know what the `txn_global->rwlock` is guarding.

I wish I could answer your question in detail, but I can't. I'm not entirely sure what the point of `txn_global.rwlock` is either -- that's one of the questions I'm asking in WT-9354, and absent that answer it may be the case that we need `rwlock` there.

But I am as confident as I can be that `rwlock` doesn't matter for the purposes of choosing `session_cnt`, it's going to grow/shrink entirely independently of `rwlock`.